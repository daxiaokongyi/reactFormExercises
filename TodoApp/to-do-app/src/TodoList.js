import React, {useState} from 'react';
import NewTodoForm from './NewTodoForm';
import Todo from './Todo';

const TodoList = () => {
    // set useState for Todos
    const [todos, setTodos] = useState([]);

    // add a new todo
    const addTodo = todo => {
        setTodos(todos => [...todos, todo]);
    }

    // remove a todo
    const removeTodo = (id) => {
        setTodos(todos => todos.filter(todo => todo.id !== id));
    }

    // update a todo
    const updateTodo = (id, updatedTodo) => {
        setTodos(
            todos => todos.map(
                todo => todo.id === id ? {...todo, todo: updatedTodo} : todo
            )
        )
    }

    const renderTodo = todos.map(todo => 
        <Todo 
            key = {todo.id}
            todo = {todo.todo} 
            remove = {removeTodo} 
            update = {updateTodo}
            id = {todo.id}
        />
    );

    return (
        <div>
            {renderTodo}
            <NewTodoForm addTodo = {addTodo} />
        </div>
    )
}

export default TodoList;