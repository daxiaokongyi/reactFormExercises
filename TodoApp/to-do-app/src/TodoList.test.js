import React from 'react';
import {render, fireEvnt, fireEvent} from "@testing-library/react";
import TodoList from './TodoList';

const addTodo = (todoList, task="add a todo test") => {
    const todoInput = todoList.getByLabelText("New Todo:");
    fireEvent.change(todoInput, {target: {value:task}});
    const submitBtn = todoList.getByText("Add a Todo");
    fireEvent.click(submitBtn);
}

// smoke test
it("renders without crashing", function() {
  render(<TodoList />);
});

// snapshot test
it("matches snapshot", function() {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

// test if it can add a todo
it("can add a todo", () => {
    const todoList = render(<TodoList/>);
    addTodo(todoList);

    // expect form to clear and todo to be on the page
    expect(todoList.getByLabelText("New Todo:")).toHaveValue("");
    expect(todoList.getByText("add a todo test")).toBeInTheDocument();
    expect(todoList.getByText("Edit")).toBeInTheDocument();
    expect(todoList.getByText("X")).toBeInTheDocument();
})

// test if it can edit a todo
it("can edit a todo", () => {
    const todoList = render(<TodoList/>);
    addTodo(todoList);

    const editBtn = todoList.getByText("Edit");
    fireEvent.click(editBtn);
    const editInput = todoList.getByDisplayValue("add a todo test");
    fireEvent.change(editInput, {target: {value: "wow~~~"}});
    const updateBtn = todoList.getByText("Update");
    fireEvent.click(updateBtn);

    expect(todoList.getByText("wow~~~")).toBeInTheDocument();
    expect(todoList.queryByText("add a todo test")).not.toBeInTheDocument();
})

// test if it can remove a todo
it("can remove a todo", () => {
    const todoList = render(<TodoList/>);
    addTodo(todoList);

    const removeBtn = todoList.getByText("X");
    fireEvent.click(removeBtn);

    expect(todoList.queryByText("add a todo test")).not.toBeInTheDocument();
})