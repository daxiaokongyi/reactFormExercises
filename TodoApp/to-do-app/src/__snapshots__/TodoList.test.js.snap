// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <label
          for="todo"
        >
          New Todo: 
        </label>
        <input
          id="todo"
          name="todo"
          type="text"
          value=""
        />
        <button>
          Add a Todo
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
